lib/service/service.go

```diff
@@ -1998,11 +2013,101 @@ func (process *TeleportProcess) initUploaderService(accessPoint auth.AccessPoint
 	return nil
 }
 
+// initMetricsService starts the metrics service currently serving metrics for
+// prometheus consumption
+func (process *TeleportProcess) initMetricsService() error {
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+
+	log := process.log.WithFields(logrus.Fields{
+		trace.Component: teleport.Component(teleport.ComponentMetrics, process.id),
+	})
+
+	listener, err := process.importOrCreateListener(listenerMetrics, process.Config.Metrics.ListenAddr.Addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	warnOnErr(process.closeImportedDescriptors(teleport.ComponentMetrics), log)
+
+	tlsConfig := &tls.Config{}
+	if process.Config.Metrics.MTLS {
+		for _, pair := range process.Config.Metrics.KeyPairs {
+			certificate, err := tls.LoadX509KeyPair(pair.Certificate, pair.PrivateKey)
+			if err != nil {
+				return trace.Wrap(err, "failed to read keypair")
+			}
+			tlsConfig.Certificates = append(tlsConfig.Certificates, certificate)
+		}
+
+		if len(tlsConfig.Certificates) == 0 {
+			return trace.Wrap(nil, "no keypairs were provided for the metrics service with mtls enabled")
+		}
+
+		pool := x509.NewCertPool()
+		for _, caCertPath := range process.Config.Metrics.CACerts {
+			caCert, err := ioutil.ReadFile(caCertPath)
+			if err != nil {
+				return trace.Wrap(err, "failed to read prometheus ca cert")
+			}
+
+			if !pool.AppendCertsFromPEM(caCert) {
+				return trace.Wrap(nil, "failed to parse prometheus ca cert")
+			}
+		}
+
+		if len(pool.Subjects()) == 0 {
+			return trace.Wrap(nil, "no prometheus ca certs were provided for the metrics service with mtls enabled")
+		}
+
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+		tlsConfig.ClientCAs = pool
+		tlsConfig.BuildNameToCertificate()
+
+		listener = tls.NewListener(listener, tlsConfig)
+	}
+
+	server := &http.Server{
+		Handler:           mux,
+		ReadHeaderTimeout: defaults.DefaultDialTimeout,
+		ErrorLog:          utils.NewStdlogger(log.Error, teleport.ComponentMetrics),
+		TLSConfig:         tlsConfig,
+	}
+
+	log.Infof("Starting metrics service on %v.", process.Config.Metrics.ListenAddr.Addr)
+
+	process.RegisterFunc("metrics.service", func() error {
+		err := server.Serve(listener)
```

Comment:
```suggestion
	}

	server := &http.Server{
		Handler:           mux,
		ReadHeaderTimeout: defaults.DefaultDialTimeout,
		ErrorLog:          utils.NewStdlogger(log.Error, teleport.ComponentMetrics),
		TLSConfig:         tlsConfig,
	}

	log.Infof("Starting metrics service on %v.", process.Config.Metrics.ListenAddr.Addr)

	process.RegisterFunc("metrics.service", func() error {
		err := server.ServeTLS(listener, "", "")
```

Reply:
I think that doesn't work if the `mtls = off` in the config. You'd be calling `server.ServerTLS` with an empty tls config and empty strings for cert  and key.